src/pyclass.rs

```diff
@@ -0,0 +1,557 @@
+    }
+    unsafe fn py_drop(&mut self, py: Python) {
+        ManuallyDrop::drop(&mut self.pyclass);
+        self.dict.clear_dict(py);
+        self.weakref.clear_weakrefs(self.as_ptr(), py);
+        self.ob_base.py_drop(py);
+    }
+    unsafe fn py_init(&mut self, value: T) {
+        self.pyclass = ManuallyDrop::new(value);
+    }
+}
+
+impl<T: PyClass> PyObjectSizedLayout<T> for PyClassShell<T> {}
+
+impl<T: PyClass> AsPyPointer for PyClassShell<T> {
+    fn as_ptr(&self) -> *mut ffi::PyObject {
+        (self as *const _) as *mut _
+    }
+}
+
+impl<T: PyClass> std::ops::Deref for PyClassShell<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.pyclass.deref()
+    }
+}
+
+impl<T: PyClass> std::ops::DerefMut for PyClassShell<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.pyclass.deref_mut()
+    }
+}
+
+impl<T: PyClass> ToPyObject for &PyClassShell<T> {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
+    }
+}
+
+impl<T: PyClass> ToPyObject for &mut PyClassShell<T> {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
+    }
+}
+
+unsafe impl<'p, T> FromPyPointer<'p> for &'p PyClassShell<T>
+where
+    T: PyClass,
+{
+    unsafe fn from_owned_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr).map(|p| &*(gil::register_owned(py, p).as_ptr() as *const PyClassShell<T>))
+    }
+    unsafe fn from_borrowed_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr)
+            .map(|p| &*(gil::register_borrowed(py, p).as_ptr() as *const PyClassShell<T>))
+    }
+}
+
+unsafe impl<'p, T> FromPyPointer<'p> for &'p mut PyClassShell<T>
+where
+    T: PyClass,
+{
+    unsafe fn from_owned_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr).map(|p| {
+            &mut *(gil::register_owned(py, p).as_ptr() as *const PyClassShell<T> as *mut _)
+        })
+    }
+    unsafe fn from_borrowed_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr).map(|p| {
+            &mut *(gil::register_borrowed(py, p).as_ptr() as *const PyClassShell<T> as *mut _)
+        })
+    }
+}
+
+/// An initializer for `PyClassShell<T>`.
+pub struct PyClassInitializer<T: PyTypeInfo> {
+    init: Option<T>,
+    super_init: Option<*mut PyClassInitializer<T::BaseType>>,
+}
+
+impl<T: PyTypeInfo> PyClassInitializer<T> {
+    pub fn from_value(value: T) -> Self {
+        PyClassInitializer {
+            init: Some(value),
+            super_init: None,
+        }
+    }
+
+    pub fn new() -> Self {
+        PyClassInitializer {
+            init: None,
+            super_init: None,
+        }
+    }
+
+    #[must_use]
+    #[doc(hiddden)]
+    pub fn init_class(self, shell: &mut T::ConcreteLayout) -> PyResult<()> {
```

Comment:
shouldn't this be `unsafe`, it seems likely to be UB to invoke twice on the same `shell` object

Reply:
Hmm :thinking: ... , I don't think it's UB.
`init` just do `shell.pyclass = value`.

Reply:
doesn't this call the CPython init functions for the base classes? Those are likely to not be allowed to be called twice on the same object.

Reply:
No, it's a part of `tp_new` and doesn't call init function.

Reply:
ah, ok. Related, if `shell` is partially uninitialized, it would be UB for safe code to read from it, so it should have `MaybeUninitialized` and/or be passed in as `*mut T::ConcreteLayout`. Also, if later code depends on `shell` being initialized instead of being `uninit`, then that's also a good reason for `init_class` to be `unsafe`.